client/state/plugins/upload/reducer.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * Internal dependencies
+ */
+import {
+	combineReducers,
+	createReducer,
+	keyedReducer,
+} from 'state/utils';
+
+import {
+	PLUGIN_UPLOAD,
+	PLUGIN_UPLOAD_CLEAR,
+	PLUGIN_UPLOAD_COMPLETE,
+	PLUGIN_UPLOAD_ERROR,
+	PLUGIN_UPLOAD_PROGRESS,
+} from 'state/action-types';
+
+export const uploadedPluginId = keyedReducer( 'siteId', createReducer( {}, {
+	[ PLUGIN_UPLOAD ]: () => null,
+	[ PLUGIN_UPLOAD_COMPLETE ]: ( state, { pluginId } ) => pluginId,
+	[ PLUGIN_UPLOAD_CLEAR ]: () => null,
+	[ PLUGIN_UPLOAD_ERROR ]: () => null,
+} ) );
+
+export const uploadError = keyedReducer( 'siteId', createReducer( {}, {
+	[ PLUGIN_UPLOAD_ERROR ]: ( state, { error } ) => error,
+	[ PLUGIN_UPLOAD ]: () => null,
+	[ PLUGIN_UPLOAD_CLEAR ]: () => null,
+	[ PLUGIN_UPLOAD_COMPLETE ]: () => null,
```

Comment:
Yeah I'm undecided on this. These inlined functions seem easy enough to read so I think I'll stick with them.